File: sdk/reproit-backend-php/exchange.php
<?php

/*!
 * Bounded dependency-exchange records for reproit-backend-php: the request
 * the app sent and the response the dependency returned.
 *
 * PHP port of the shapes in sdk/reproit-backend-node/instrument.js. An
 * exchange is what hermetic replay serves, so responses are captured
 * verbatim up to a fixed inline budget; an over-budget body keeps only
 * provable identity (byte count + sha256) and is marked truncated, and
 * replay fails closed on it with a named reason instead of guessing.
 */

declare(strict_types=1);

namespace ReproitBackend;

/** Inline body budget per exchange side, identical to every other SDK. */
const MAX_EXCHANGE_BODY_BYTES = 8 * 1024;
/** Recorded headers are capped to keep events bounded. */
const MAX_EXCHANGE_HEADERS = 32;
/** Rows recorded per database result; beyond it the result is truncated. */
const MAX_DB_ROWS = 64;
/**
 * Stream chunk boundaries recorded per exchange (SSE / chunked responses,
 * the LLM streaming shape). Beyond it the boundaries are marked truncated
 * and replay fails closed rather than serve a wrong stream shape.
 */
const MAX_STREAM_CHUNKS = 128;

/**
 * Bound one exchange body. Declared JSON is decoded so structural redaction
 * sees fields rather than text. A BodyCollector that overflowed already
 * reduced the body to provable identity (byte count + sha256); that identity
 * array passes through untouched.
 */
function bounded_body(mixed $body, string $contentType): array
{
    if (\is_array($body) && ($body['truncated'] ?? false) === true) {
        return $body;
    }
    if (!\is_string($body) || $body === '') {
        return [];
    }
    if (\strlen($body) > MAX_EXCHANGE_BODY_BYTES) {
        return [
            'bodyBytes' => \strlen($body),
            'bodySha256' => hash('sha256', $body),
            'truncated' => true,
        ];
    }
    if (str_contains($contentType, 'application/json')) {
        $decoded = json_decode($body, true);
        if (json_last_error() === JSON_ERROR_NONE) {
            return ['body' => $decoded];
        }
    }
    return ['body' => $body];
}

/** @param array<string, mixed> $headers */
function bounded_headers(array $headers): array
{
    $lowered = [];
    foreach ($headers as $name => $value) {
        $lowered[strtolower((string) $name)] = \is_array($value)
            ? implode(', ', array_map('strval', $value))
            : (string) $value;
    }
    // Sort BEFORE the cap. Capping arrival order records a different subset
    // whenever the caller's header order shifts, so two runs of one request
    // disagree and the capsule stops matching.
    ksort($lowered, SORT_STRING);
    $bounded = \array_slice($lowered, 0, MAX_EXCHANGE_HEADERS, true);
    return $bounded === [] ? [] : ['headers' => $bounded];
}

/**
 * Collect a stream's chunks up to one byte past the inline budget; enough to
 * know the true size class without holding unbounded memory. The sha256 runs
 * over EVERY byte so truncated identity stays provable. Chunk boundaries are
 * recorded as observed byte lengths, bounded by MAX_STREAM_CHUNKS;
 * boundaries past the cap are counted, never guessed.
 *
 * PHP port of the Node reference's bodyCollector (instrument.js).
 */
final class BodyCollector
{
    private array $chunks = [];
    private array $boundaries = [];
    private int $bytes = 0;
    private int $droppedBoundaries = 0;
    private \HashContext $hash;

    public function __construct()
    {
        $this->hash = hash_init('sha256');
    }

    public function push(string $chunk): void
    {
        $this->bytes += \strlen($chunk);
        hash_update($this->hash, $chunk);
        if (\count($this->boundaries) < MAX_STREAM_CHUNKS) {
            $this->boundaries[] = \strlen($chunk);
        } else {
            $this->droppedBoundaries += 1;
        }
        if ($this->bytes <= MAX_EXCHANGE_BODY_BYTES) {
            $this->chunks[] = $chunk;
        }
    }

    /**
     * The collected body: null when empty, provable identity (an array
     * bounded_body passes through) when over budget, the raw bytes otherwise.
     */
    public function result(): string|array|null
    {
        if ($this->bytes === 0) {
            return null;
        }
        if ($this->bytes > MAX_EXCHANGE_BODY_BYTES) {
            return [
                'bodyBytes' => $this->bytes,
                'bodySha256' => hash_final(clone $this->hash),
                'truncated' => true,
            ];
        }
        return implode('', $this->chunks);
    }

    /**
     * Chunk boundaries as observed byte lengths. Recorded when the response
     * is a stream (SSE always; anything else only when it actually arrived
     * in more than one chunk, since a single-chunk body replays identically
     * without them).
     */
    public function stream(bool $isEventStream): ?array
    {
        if ($this->boundaries === []) {
            return null;
        }
        if (!$isEventStream && \count($this->boundaries) < 2 && $this->droppedBoundaries === 0) {
            return null;
        }
        if ($this->droppedBoundaries > 0) {
            return ['chunks' => $this->boundaries, 'truncated' => true];
        }
        return ['chunks' => $this->boundaries];
    }
}

/**
 * @param array{method: string, url: string, headers: array, body: mixed,
 *              contentType: string} $request
 * @param array{status: int, headers: array, body: mixed,
 *              contentType: string, stream?: ?array} $response
 */
function http_exchange(array $request, array $response): array
{
    $responseBody = bounded_body($response['body'], $response['contentType']);
    $encoded = array_merge(
        ['status' => $response['status']],
        bounded_headers($response['headers']),
        $responseBody
    );
    // Stream shape (SSE / chunked): observed chunk boundaries, so the whole
    // stream is ONE logical exchange and replay can re-serve it chunk for
    // chunk. A truncated inline body already fails closed, so boundaries are
    // only kept for bodies recorded verbatim.
    $stream = $response['stream'] ?? null;
    if (\is_array($stream) && ($responseBody['truncated'] ?? false) !== true) {
        $encoded['stream'] = $stream;
    }
    return [
        'protocol' => 'http',
        'request' => array_merge(
            ['method' => $request['method'], 'url' => $request['url']],
            bounded_headers($request['headers']),
            bounded_body($request['body'], $request['contentType'])
        ),
        'response' => $encoded,
    ];
}

function db_exchange(string $text, ?array $values, array $outcome): array
{
    $request = ['text' => $text];
    if ($values !== null && $values !== []) {
        $request['values'] = array_values($values);
    }
    return ['protocol' => 'db', 'request' => $request, 'response' => $outcome];
}

/** Normalize a driver result into the recorded response shape. */
function db_outcome(mixed $result): array
{
    if (!\is_array($result)) {
        return ['rowCount' => 0];
    }
    $rows = \is_array($result['rows'] ?? null) ? array_values($result['rows']) : [];
    $count = $result['rowCount'] ?? null;
    $outcome = [
        'command' => isset($result['command']) ? (string) $result['command'] : null,
        'rowCount' => \is_int($count) ? $count : \count($rows),
        'rows' => \array_slice($rows, 0, MAX_DB_ROWS),
    ];
    if (\count($rows) > MAX_DB_ROWS) {
        $outcome['truncated'] = true;
    }
    return $outcome;
}

function db_error(\Throwable $error): array
{
    $code = $error->getCode();
    return ['error' => [
        'message' => $error->getMessage(),
        'code' => $code === 0 ? null : (string) $code,
    ]];
}

/**
 * Effect kind for a statement: reads stay reads so state oracles keep their
 * meaning; everything else is a write.
 */
function statement_effect_kind(string $text): string
{
    $verb = strtoupper(substr(ltrim($text), 0, 8));
    return str_starts_with($verb, 'SELECT') || str_starts_with($verb, 'SHOW')
        ? 'read'
        : 'write';
}

File: sdk/reproit-backend-rb/lib/reproit_backend_rb/exchange.rb
# Bounded dependency-exchange records: the request the app sent and the
# response the dependency returned.
#
# Ruby port of the shapes in sdk/reproit-backend-node/instrument.js. An
# exchange is what hermetic replay serves, so responses are captured verbatim
# up to a fixed inline budget; an over-budget body keeps only provable
# identity (byte count + sha256) and is marked truncated, and replay fails
# closed on it with a named reason instead of guessing.

require "digest"
require "json"

module ReproitBackendRb
  module Exchange
    # Inline body budget per exchange side, byte-identical to the Node and
    # Rust SDKs so one replay engine consumes every platform's captures.
    MAX_EXCHANGE_BODY_BYTES = 8 * 1024
    # Recorded headers are capped to keep events bounded.
    MAX_EXCHANGE_HEADERS = 32
    # Rows recorded per database result; beyond it the result is truncated.
    MAX_DB_ROWS = 64
    # Stream chunk boundaries recorded per exchange (SSE / chunked responses,
    # the LLM streaming shape). Beyond it the boundaries are marked truncated
    # and replay fails closed rather than serve a wrong stream shape.
    MAX_STREAM_CHUNKS = 128

    module_function

    # Bound one exchange body. Declared JSON is parsed so structural
    # redaction sees fields rather than text.
    def bounded_body(body, content_type)
      return {} if body.nil?
      # A BodyCollector that overflowed already reduced the body to provable
      # identity; pass it through instead of stringifying the identity hash.
      return body if body.is_a?(Hash) && body["truncated"] == true
      bytes = body.to_s.dup.force_encoding(Encoding::BINARY)
      return {} if bytes.empty?
      if bytes.bytesize > MAX_EXCHANGE_BODY_BYTES
        return {
          "bodyBytes" => bytes.bytesize,
          "bodySha256" => Digest::SHA256.hexdigest(bytes),
          "truncated" => true,
        }
      end
      text = bytes.dup.force_encoding(Encoding::UTF_8)
      if content_type.to_s.include?("application/json")
        begin
          return { "body" => JSON.parse(text) }
        rescue JSON::ParserError, EncodingError
          # Declared JSON that does not parse is recorded as text below.
        end
      end
      { "body" => text.valid_encoding? ? text : bytes.unpack1("H*") }
    end

    # Sorted BEFORE the cap. Capping arrival order records a different subset
    # whenever the caller's header order shifts, so two runs of one request
    # disagree and the capsule stops matching.
    def bounded_headers(headers)
      entries = (headers || {}).map do |name, value|
        [name.to_s.downcase, value.is_a?(Array) ? value.join(", ") : value.to_s]
      end
      entries = entries.sort_by(&:first).first(MAX_EXCHANGE_HEADERS)
      entries.empty? ? {} : { "headers" => entries.to_h }
    end

    # `request`/`response` are plain hashes of already-collected parts.
    def http(request, response)
      response_body = bounded_body(response[:body], response[:content_type])
      response_value = { "status" => response[:status] }
        .merge(bounded_headers(response[:headers]))
        .merge(response_body)
      # Stream shape (SSE / chunked): observed chunk boundaries, so the
      # whole stream is ONE logical exchange and replay can re-serve it
      # chunk for chunk. A truncated inline body already fails closed, so
      # boundaries are only kept for bodies recorded verbatim.
      stream = response[:stream]
      if stream && response_body["truncated"] != true
        response_value["stream"] = stream
      end
      {
        "protocol" => "http",
        "request" => {
          "method" => request[:method],
          "url" => request[:url],
        }.merge(bounded_headers(request[:headers]))
          .merge(bounded_body(request[:body], request[:content_type])),
        "response" => response_value,
      }
    end

    # Collect a stream's chunks up to one byte past the inline budget;
    # enough to know the true size class without holding unbounded memory.
    # The sha256 runs over EVERY byte so truncated identity stays provable.
    # Chunk boundaries are recorded as observed byte lengths, bounded by
    # MAX_STREAM_CHUNKS; boundaries past the cap are counted, never guessed.
    class BodyCollector
      def initialize
        @chunks = []
        @boundaries = []
        @bytes = 0
        @dropped_boundaries = 0
        @digest = Digest::SHA256.new
      end

      def push(chunk)
        raw = chunk.to_s.dup.force_encoding(Encoding::BINARY)
        @bytes += raw.bytesize
        @digest.update(raw)
        if @boundaries.length < MAX_STREAM_CHUNKS
          @boundaries << raw.bytesize
        else
          @dropped_boundaries += 1
        end
        @chunks << raw if @bytes <= MAX_EXCHANGE_BODY_BYTES
        nil
      end

      # The collected body: nil when empty, provable identity when over
      # budget, the raw bytes otherwise.
      def result
        return nil if @bytes.zero?
        if @bytes > MAX_EXCHANGE_BODY_BYTES
          return {
            "bodyBytes" => @bytes,
            "bodySha256" => @digest.hexdigest,
            "truncated" => true,
          }
        end
        @chunks.join
      end

      def truncated?
        @bytes > MAX_EXCHANGE_BODY_BYTES
      end

      # Chunk boundaries as observed byte lengths. Recorded when the
      # response is a stream (SSE always; anything else only when it
      # actually arrived in more than one chunk, since a single-chunk body
      # replays identically without them).
      def stream(is_event_stream)
        return nil if @boundaries.empty?
        if !is_event_stream && @boundaries.length < 2 && @dropped_boundaries.zero?
          return nil
        end
        return { "chunks" => @boundaries.dup, "truncated" => true } if @dropped_boundaries.positive?
        { "chunks" => @boundaries.dup }
      end
    end

    def db(text, values, outcome)
      request = { "text" => text.to_s }
      request["values"] = values if values.is_a?(Array) && !values.empty?
      { "protocol" => "db", "request" => request, "response" => outcome }
    end

    # Normalize a driver result into the recorded response shape. `rows` is
    # any array of row hashes; anything else records as a bare row count.
    def db_outcome(result)
      return { "rowCount" => 0 } unless result.is_a?(Hash)
      rows = result["rows"] || result[:rows]
      rows = [] unless rows.is_a?(Array)
      command = result["command"] || result[:command]
      count = result["rowCount"] || result[:rowCount] || result[:row_count]
      outcome = {
        "command" => command.nil? ? nil : command.to_s,
        "rowCount" => count.is_a?(Integer) ? count : rows.length,
        "rows" => rows.first(MAX_DB_ROWS),
      }
      outcome["truncated"] = true if rows.length > MAX_DB_ROWS
      outcome
    end

    def db_error(error)
      code = error.respond_to?(:code) ? error.code : nil
      { "error" => { "message" => error.message.to_s, "code" => code.nil? ? nil : code.to_s } }
    end

    # Effect kind for a statement: reads stay reads so state oracles keep
    # their meaning; everything else is a write.
    def statement_effect_kind(text)
      verb = text.to_s.lstrip[0, 8].to_s.upcase
      verb.start_with?("SELECT", "SHOW") ? "read" : "write"
    end
  end
end
